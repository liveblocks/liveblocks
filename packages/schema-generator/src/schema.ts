import { mergeInferredTypes, InferredType, isAtomic } from ".";
import { InferredObjectType, inferredObjectTypeToAst } from "./object";
import { InferredTypeReference } from "./typeReference";
import { invariant } from "./utils/invariant";
import { BidirectionalMap } from "./utils/bidirectionalMap";
import { AST } from "@liveblocks/schema";

type RootTypes = Set<InferredObjectType>;
type RootReferences = Map<InferredObjectType, Set<InferredTypeReference>>;
type RootNames = BidirectionalMap<string, InferredObjectType>;

export type InferredSchema = {
  rootTypes: RootTypes;
  rootTypeReferences: RootReferences;
  rootNames: RootNames;
};

// Deeply iterate over a inferred type, yielding the type and optionally the type reference
// object referencing it.
// If yieldCircularReferences is true, the same type can be yielded multiple times
// if another type circularly references it in order to yield all references to that type.
function* iterateInferredTypesDeep(
  root: InferredType,
  yieldCircularReferences = true
): Generator<[InferredType, InferredTypeReference | undefined]> {
  const visited = new Set<InferredType>();

  function* iterate(
    root: InferredType,
    referenceObj: InferredTypeReference | undefined = undefined
  ): Generator<[InferredType, InferredTypeReference | undefined]> {
    if (visited.has(root)) {
      if (yieldCircularReferences && referenceObj) {
        yield [root, referenceObj];
      }

      return;
    }

    visited.add(root);

    yield [root, referenceObj];
    switch (root.type) {
      case "Object":
        for (const field of root.fields.values()) {
          yield* iterate(field.value, field);
        }
        return;
    }
  }

  yield* iterate(root);
}

function inferRootTypes(
  inferred: InferredObjectType
): Pick<InferredSchema, "rootTypes" | "rootTypeReferences"> {
  const rootTypes: RootTypes = new Set();
  const rootTypeReferences: RootReferences = new Map();

  for (const [type, referenceObj] of iterateInferredTypesDeep(inferred)) {
    if (type.type === "Object") {
      rootTypes.add(type);

      if (referenceObj) {
        const references = rootTypeReferences.get(type) ?? new Set();
        references.add(referenceObj);
        rootTypeReferences.set(type, references);
      }
    }
  }

  return { rootTypes, rootTypeReferences };
}

function replaceRootType(
  schema: InferredSchema,
  oldType: InferredObjectType,
  newType: InferredObjectType
) {
  invariant(schema.rootTypes.has(oldType), "Old type not part of the schema");

  schema.rootTypes.delete(oldType);
  schema.rootTypes.add(newType);
  schema.rootNames.deleteValue(oldType);

  // Re-wire all references to the old type to the new type
  const oldTypeReferences = schema.rootTypeReferences.get(oldType);
  if (oldTypeReferences) {
    const newReferences = schema.rootTypeReferences.get(newType) ?? new Set();
    for (const reference of oldTypeReferences) {
      reference.value = newType;
      newReferences.add(reference);
    }
    schema.rootTypeReferences.set(newType, oldTypeReferences);
  }
}

function assignNameOrMerge(schema: InferredSchema, type: InferredObjectType) {
  invariant(schema.rootTypes.has(type), "Root type not part of the schema");
  invariant(!schema.rootNames.hasValue(type), "Root type already has a name");

  const orderedNames = Array.from(type.names.entries())
    .sort(([, scoreA], [, scoreB]) => scoreB - scoreA)
    .map(([name]) => name);

  // TODO: Be smarter about when to merge types, and when to rename them.
  // Might also be worth a try to rename both types e.g. "Data" conflicts with "Data" ->
  // rename them to "UserData" and "ShapeData" instead of keeping "Data" and adding "UserData".
  for (const name of orderedNames) {
    const existingType = schema.rootNames.get(name);
    if (!existingType) {
      schema.rootNames.set(name, type);
      return;
    }

    const combined = mergeInferredTypes(existingType, type) as
      | InferredObjectType
      | undefined;

    if (combined) {
      replaceRootType(schema, existingType, combined);
      replaceRootType(schema, type, combined);
      schema.rootNames.set(name, combined);
      return;
    }
  }

  // Try Name2, Name3, etc. until we find a name that doesn't conflict.
  let suffix = 2;
  while (true) {
    const name = `${orderedNames[0]}${suffix}`;
    if (!schema.rootNames.has(name)) {
      schema.rootNames.set(name, type);
      return;
    }

    suffix++;
  }
}

export function inferSchema(inferredStorage: InferredObjectType) {
  const rootTypes = inferRootTypes(inferredStorage);
  const schema: InferredSchema = {
    ...rootTypes,
    rootNames: new BidirectionalMap(),
  };

  const toAssign = Array.from(schema.rootTypes.values()).sort((a, b) => {
    const aIsAtomic = isAtomic(a);
    const bIsAtomic = isAtomic(b);
    if (aIsAtomic && bIsAtomic) {
      return 0;
    }
    if (aIsAtomic || bIsAtomic) {
      return aIsAtomic ? -1 : 1;
    }

    // TODO: Assign names to root types with fewer aliases first?
    return Math.max(...a.names.values()) - Math.max(...b.names.values());
  });

  toAssign.forEach((type) => assignNameOrMerge(schema, type));
  return schema;
}

export function inferredSchemaToAst(schema: InferredSchema): AST.Document {
  return {
    _kind: "Document",
    definitions: Array.from(schema.rootTypes.values()).map((rootType) =>
      inferredObjectTypeToAst(rootType, schema)
    ),
    range: [0, 0],
  };
}
