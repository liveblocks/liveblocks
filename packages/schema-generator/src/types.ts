import type { Json } from "@liveblocks/core";

import type { InferredObjectType } from "./object";
import { isInferredObjectType } from "./object";
import type { InferredScalarType } from "./scalar";
import type { PartialBy } from "./utils/types";

export type PlainLsonFields = Record<string, PlainLson>;
export type PlainLsonObject = {
  liveblocksType: "LiveObject";
  data: PlainLsonFields;
};

export type PlainLsonMap = {
  liveblocksType: "LiveMap";
  data: PlainLsonFields;
};

export type PlainLsonList = {
  liveblocksType: "LiveList";
  data: PlainLson[];
};

export type { Json, JsonObject, JsonScalar } from "@liveblocks/core";

export type PlainLson =
  | PlainLsonObject
  | PlainLsonMap
  | PlainLsonList

  // Any "normal" Json value, as long as it's not an object with
  // a `liveblocksType` field :)
  | Json;

type FieldChildContext = {
  parent: PartialBy<InferredObjectType, "fields">;
  field: string;
};

export type ChildContext = FieldChildContext & { json: boolean }; // TODO: Expand for union, list, ...

export type InferredType = InferredScalarType | InferredObjectType;

export function isAtomic(type: InferredType): boolean {
  return isInferredObjectType(type) && type.atomic;
}
