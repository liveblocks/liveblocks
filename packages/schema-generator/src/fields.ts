import { AST } from "@liveblocks/schema";
import { InferredSchema } from "./schema";
import {
  combineTypeReferences,
  InferredTypeReference,
  inferredTypeReferenceToAst,
  inferTypeReference,
} from "./typeReference";
import { ChildContext, JsonObject, PlainLsonFields } from "./types";
import { invariant } from "./utils/invariant";
import { isNotUndefined } from "./utils/typeGuards";

export type InferredFields = Record<string, InferredTypeReference>;

export function inferLsonFields(
  fields: PlainLsonFields | JsonObject,
  ctx: Omit<ChildContext, "field">
): InferredFields {
  const fieldEntries = Object.entries(fields)
    .map(([key, value]) => {
      if (value === undefined) {
        return undefined;
      }

      return [key, inferTypeReference(value, { ...ctx, field: key })] as const;
    })
    .filter(isNotUndefined);

  return Object.fromEntries(fieldEntries);
}

export function combineInferredFields(
  a: InferredFields,
  b: InferredFields
): InferredFields | undefined {
  const keys = new Set([...Object.keys(a), ...Object.keys(b)]);

  const mergedFields: InferredFields = {};
  for (const key of keys) {
    const valueA = a[key];
    const valueB = b[key];

    if (!valueA || !valueB) {
      const mergedValue = valueA ?? valueB;

      // Should never happen
      invariant(isNotUndefined(mergedValue));

      mergedValue[key] = { ...mergedValue, optional: true };
      continue;
    }

    const mergedValue = combineTypeReferences(valueA, valueB);
    if (!mergedValue) {
      return undefined;
    }

    mergedFields[key] = mergedValue;
  }

  return mergedFields;
}

export function inferredFieldsToAst(
  fields: InferredFields,
  schema: InferredSchema
): AST.FieldDef[] {
  return Object.entries(fields).map(([name, value]) => ({
    _kind: "FieldDef",
    name: { _kind: "Identifier", name, range: [0, 0] },
    type: inferredTypeReferenceToAst(value, schema),
    optional: value.optional,
    range: [0, 0],
  }));
}
