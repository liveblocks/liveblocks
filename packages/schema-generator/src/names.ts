import { singular, ucFirst, words } from "./utils/strings";
import { ChildContext } from "./types";

export type ScoredNames = Map<string, number>;

export function mergeScoredNames(a: ScoredNames, b: ScoredNames): ScoredNames {
  const merged = new Map();

  const keys = new Set([...a.keys(), ...b.keys()]);
  for (const key of keys) {
    const aScore = a.get(key) ?? 0;
    const bScore = b.get(key) ?? 0;
    merged.set(key, aScore + bScore);
  }

  return merged;
}

// "gameFields" -> "GameField", "__TEST_DATA__" -> "TestData" etc.
function fieldToTypeName(fieldName: string): string {
  return words(fieldName)
    .map((word) => singular(ucFirst(word)))
    .join("");
}

export function generateNames({ field, parent }: ChildContext): ScoredNames {
  const typeName = fieldToTypeName(field);

  const names: ScoredNames = new Map();
  names.set(typeName, 1);

  for (const [name, score] of parent.names.entries()) {
    names.set(`${name}${typeName}`, score * 0.5);
  }

  return names;
}
