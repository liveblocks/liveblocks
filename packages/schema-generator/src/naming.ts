import type { ChildContext } from "./inference";
import { singular, ucFirst, words } from "./utils/strings";

export type ScoredNames = Record<string, number>;

export function mergeScoredNames(a: ScoredNames, b: ScoredNames): ScoredNames {
  const keys = new Set([...Object.keys(a), ...Object.keys(b)]);
  const mergedEntries = Array.from(keys.values(), (key) => [
    key,
    (a[key] ?? 0) + (b[key] ?? 0),
  ]);
  return Object.fromEntries(mergedEntries);
}

// "gameFields" -> "GameField", "__TEST_DATA__" -> "TestData" etc.
function fieldToTypeName(fieldName: string): string {
  return words(fieldName)
    .map((word) => singular(ucFirst(word)))
    .join("");
}

export function generateNames({ field, parent }: ChildContext): ScoredNames {
  const typeName = fieldToTypeName(field);

  const aliasEntries = Object.entries(parent.names)
    .map(([name, score]) => [
      // Data -> ScoreData, ScoreData -> GameScoreData etc.
      `${name}${typeName}`,
      score * 0.5,
    ])
    // Only keep aliases with a score of at least 0.25 (= a depth of at most 3)
    // to avoid overly long names
    .filter(([, score]) => score >= 0.25);

  return mergeScoredNames({ [typeName]: 1 }, Object.fromEntries(aliasEntries));
}
