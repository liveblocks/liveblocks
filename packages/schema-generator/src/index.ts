import { inferLsonFields } from "./fields";
import {
  inferObjectType,
  InferredObjectType,
  isInferredObjectType,
  mergeInferredObjectTypes,
} from "./object";
import {
  inferScalarType,
  isInferredScalarType,
  mergeInferredScalarTypes,
} from "./scalar";
import type {
  ChildContext,
  InferredType,
  JsonScalar,
  PlainLson,
  PlainLsonObject,
} from "./types";
import type { PartialBy } from "./utils/types";

export function inferStorageType(value: PlainLsonObject): InferredObjectType {
  const storage: PartialBy<InferredObjectType, "fields"> = {
    names: { Storage: 1 },
    type: "Object",
    live: true,
    atomic: true,
  };

  const fields = inferLsonFields(value.data, {
    json: false,
    parent: storage,
  });
  storage.fields = fields;

  return storage as InferredObjectType;
}

export function inferType(value: PlainLson, ctx: ChildContext): InferredType {
  if (typeof value !== "object" || value === null || value === undefined) {
    return inferScalarType(value as JsonScalar, ctx);
  }

  if (Array.isArray(value)) {
    throw new Error("Not implemented");
  }

  // If we are in a json all objects are plain struct types even if they
  // have a type property
  if (!ctx.json && "liveblocksType" in value) {
    if (value.liveblocksType === "LiveObject") {
      return inferObjectType(value as PlainLsonObject, ctx);
    }

    if (value.liveblocksType === "LiveList") {
      throw new Error("Not implemented");
    }

    if (value.liveblocksType === "LiveMap") {
      throw new Error("Not implemented");
    }
  }

  return inferObjectType(value, ctx);
}

export function mergeInferredTypes(
  a: InferredType,
  b: InferredType
): InferredType | undefined {
  if (isInferredScalarType(a) && isInferredScalarType(b)) {
    return mergeInferredScalarTypes(a, b);
  }

  if (isInferredObjectType(a) && isInferredObjectType(b)) {
    return mergeInferredObjectTypes(a, b);
  }

  // TODO: Add missing types
  return undefined;
}
