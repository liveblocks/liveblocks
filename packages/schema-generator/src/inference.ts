import { inferLsonFields } from "./fields";
import type { InferredObjectType } from "./object";
import {
  inferObjectType,
  isInferredObjectType,
  mergeInferredObjectTypes,
} from "./object";
import type { JsonScalar, PlainLson, PlainLsonObject } from "./plainLson";
import type { InferredScalarType } from "./scalar";
import {
  inferScalarType,
  isInferredScalarType,
  mergeInferredScalarTypes,
} from "./scalar";
import type { PartialBy } from "./utils/types";

type FieldChildContext = {
  parent: PartialBy<InferredObjectType, "fields">;
  field: string;
};

export type ChildContext = FieldChildContext & { json: boolean }; // TODO: Expand for union, list, ...

export type InferredType = InferredScalarType | InferredObjectType;

export function isAtomic(type: InferredType): boolean {
  return isInferredObjectType(type) && type.atomic;
}

export function inferStorageType(value: PlainLsonObject): InferredObjectType {
  const storage: PartialBy<InferredObjectType, "fields"> = {
    names: { Storage: 1 },
    type: "Object",
    live: true,
    atomic: true,
  };

  const fields = inferLsonFields(value.data, {
    json: false,
    parent: storage,
  });
  storage.fields = fields;

  return storage as InferredObjectType;
}

export function inferType(value: PlainLson, ctx: ChildContext): InferredType {
  if (typeof value !== "object" || value === null || value === undefined) {
    return inferScalarType(value as JsonScalar, ctx);
  }

  if (Array.isArray(value)) {
    throw new Error("Not implemented");
  }

  // If we are in a json all objects are plain struct types even if they
  // have a type property
  if (!ctx.json && "liveblocksType" in value) {
    if (value.liveblocksType === "LiveObject") {
      return inferObjectType(value as PlainLsonObject, ctx);
    }

    if (value.liveblocksType === "LiveList") {
      throw new Error("Not implemented");
    }

    if (value.liveblocksType === "LiveMap") {
      throw new Error("Not implemented");
    }
  }

  return inferObjectType(value, ctx);
}

export function mergeInferredTypes(
  a: InferredType,
  b: InferredType
): InferredType | undefined {
  if (isInferredScalarType(a) && isInferredScalarType(b)) {
    return mergeInferredScalarTypes(a, b);
  }

  if (isInferredObjectType(a) && isInferredObjectType(b)) {
    return mergeInferredObjectTypes(a, b);
  }

  // TODO: Add missing types
  return undefined;
}
