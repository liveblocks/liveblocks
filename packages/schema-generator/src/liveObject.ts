import { PlainLsonObject } from "./types";
import { PartialBy } from "./utils/types";
import {
  combineInferredFields,
  inferLsonFields,
  InferredFields,
  inferredFieldsToAst,
} from "./fields";
import { generateNames, mergeScoredNames, ScoredNames } from "./names";
import { ChildContext } from "./types";
import { InferredSchema } from "./schema";
import { AST } from "@liveblocks/schema";
import { invariant } from "./utils/invariant";

export type InferredLiveObjectType = {
  type: "LiveObject";
  fields: InferredFields;
  names: ScoredNames;
  atomic: boolean;
};

export function inferLiveObjectType(
  value: PlainLsonObject,
  ctx: ChildContext
): InferredLiveObjectType {
  const inferred: PartialBy<InferredLiveObjectType, "fields"> = {
    type: "LiveObject",
    atomic: false,
    names: generateNames(ctx),
  };

  inferred.fields = inferLsonFields(value.data, { ...ctx, parent: inferred });
  return inferred as InferredLiveObjectType;
}

export function combineInferredLiveObjectTypes(
  a: InferredLiveObjectType,
  b: InferredLiveObjectType
): InferredLiveObjectType | undefined {
  // Atomic types can't be combined (Mainly the root "Storage" type)
  if (a.atomic || b.atomic) {
    return undefined;
  }

  const combinedFields = combineInferredFields(a.fields, b.fields);
  if (!combinedFields) {
    return undefined;
  }

  return {
    names: mergeScoredNames(a.names, b.names),
    type: "LiveObject",
    atomic: false,
    fields: combinedFields,
  };
}

export function inferredLiveObjectTypeToAst(
  inferred: InferredLiveObjectType,
  schema: InferredSchema
): AST.ObjectTypeDefinition {
  const name = schema.rootNames.getKey(inferred);
  invariant(name != null, "Object type without assigned name");

  return {
    _kind: "ObjectTypeDefinition",
    name: { _kind: "TypeName", name, range: [0, 0] },
    fields: inferredFieldsToAst(inferred.fields, schema),
    range: [0, 0],
  };
}
