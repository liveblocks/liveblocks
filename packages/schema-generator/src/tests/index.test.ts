import { inferStorageType } from "..";
import { inferSchema, inferredSchemaToAst } from "../schema";
import { PlainLsonObject } from "../types";
import { prettify } from "@liveblocks/schema/src/prettify";

const EMPTY: PlainLsonObject = {
  liveblocksType: "LiveObject",
  data: {},
};

const BASIC_LIVE_OBJECT: PlainLsonObject = {
  liveblocksType: "LiveObject",
  data: {
    fills: {
      liveblocksType: "LiveObject",
      data: {
        mesh: "#981515",
        sole: "#808960",
        stripes: "#e3fccb",
        laces: "#c3f4bb",
      },
    },
    strokes: {
      liveblocksType: "LiveObject",
      data: {
        mesh: "#aa88ff",
        sole: "#800000",
        laces: "#00ff00",
      },
    },
  },
};

const BASIC_MERGE: PlainLsonObject = {
  liveblocksType: "LiveObject",
  data: {
    fills: {
      liveblocksType: "LiveObject",
      data: {
        mesh: { b: 1 },
        sole: "#808960",
        stripes: "#e3fccb",
        laces: "#c3f4bb",
      },
    },
    strokes: {
      liveblocksType: "LiveObject",
      data: {
        mesh: { a: 1 },
        sole: "#800000",
        laces: "#00ff00",
      },
    },
  },
};

const BASIC_UNMERGEABLE: PlainLsonObject = {
  liveblocksType: "LiveObject",
  data: {
    fills: {
      liveblocksType: "LiveObject",
      data: {
        mesh: { a: 1 },
        sole: "#808960",
        stripes: "#e3fccb",
        laces: "#c3f4bb",
      },
    },
    strokes: {
      liveblocksType: "LiveObject",
      data: {
        mesh: { a: "something" },
        sole: "#800000",
        laces: "#00ff00",
      },
    },
  },
};

describe("inferType", () => {
  const testCases = {
    EMPTY,
    BASIC_MERGE,
    BASIC_UNMERGEABLE,
    BASIC_LIVE_OBJECT,
  };

  Object.entries(testCases).forEach(([name, schema]) => {
    /*
    it(`correctly infers the "${name}" type`, () => {
      expect(inferStorageType(schema)).toMatchSnapshot();
    });

    it(`correctly infers the "${name}" schema`, () => {
      expect(inferSchema(inferStorageType(schema))).toMatchSnapshot();
    });

    it(`correctly infers the "${name}" schema ast`, () => {
      expect(
        inferredSchemaToAst(inferSchema(inferStorageType(schema)))
      ).toMatchSnapshot();
    });
    */

    it(`correctly infers the "${name}" schema`, () => {
      expect(
        prettify(inferredSchemaToAst(inferSchema(inferStorageType(schema))))
      ).toMatchSnapshot();
    });
  });
});
