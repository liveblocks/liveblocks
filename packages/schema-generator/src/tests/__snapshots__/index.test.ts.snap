// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inferType correctly infers the "BASIC_LIVE_OBJECT" schema 1`] = `
"type Storage {
  fills: LiveObject<Fill>
  strokes: LiveObject<Stroke>
}

type Fill {
  mesh: String
  sole: String
  stripes: String
  laces: String
}

type Stroke {
  mesh: String
  sole: String
  laces: String
}"
`;

exports[`inferType correctly infers the "BASIC_MERGE" schema 1`] = `
"type Storage {
  fills: LiveObject<Fill>
  strokes: LiveObject<Stroke>
}

type Fill {
  position: Position
  sole: String
  stripes: String
  laces: String
}

type Stroke {
  position: Position
  sole: String
  laces: String
}

type Position {
  x: Float
  y: Float
}"
`;

exports[`inferType correctly infers the "BASIC_UNMERGEABLE" schema 1`] = `
"type Storage {
  fills: LiveObject<Fill>
  strokes: LiveObject<Stroke>
}

type Fill {
  mesh: Mesh
  sole: String
  stripes: String
  laces: String
}

type Mesh {
  a: Int
}

type Stroke {
  mesh: StrokeMesh
  sole: String
  laces: String
}

type StrokeMesh {
  a: String
}"
`;

exports[`inferType correctly infers the "EMPTY" schema 1`] = `
"type Storage {
}"
`;
